packages/react-router/lib/router/define-route.ts

```diff
@@ -0,0 +1,512 @@
+import type { ReactNode } from "react";
+
+import type { MetaDescriptor, MetaMatch } from "../dom/ssr/routeModules";
+import type { LinkDescriptor } from "../dom/ssr/links";
+import type { Location } from "./history";
+
+// TODO: allow widest type (branded type: NOT_SET)
+
+// TODO: replace `Context` with `AppLoadContext` once `react-router` absorbs `server-runtime` (https://github.com/remix-run/react-router/pull/11669)
```

Comment:
I hope you end up renaming `AppLoadContext` because I think it was a typo. Should probably have been `AppLoaderContext`.

I really think you should use the new Data Context API for this. So context is always just an `object` with a `get/set` method. So the new types should just be `DataContext`.

[Data Context RFC](https://github.com/remix-run/react-router/discussions/11565)

So, if you want to create a custom _AppLoadContext_
```ts
// create a type
type MyAppContextType = {
  foo: string
  bar: number
}
// then create a context object
export MyAppContext = createDataContext<MyAppContextType>()

// app startup
context.set(MyAppContext, { foo: 'hello', bar: 123 })

// from loaders
context.get(MyAppContext)
```